python/lsst/pipe/tasks/drpAssociationPipe.py

```diff
@@ -225,112 +225,73 @@ def run(self,
+            # diaSources near the patch boundary can be associated.
+            # DiaObjects will be trimmed to the inner patch bbox, and any
+            # diaSources associated with dropped diaObjects will also be dropped
+            diaInPatch = self._trimToPatch(diaCat.to_pandas(), outerPatchBox, skyInfo.wcs)
 
-            nDiaSrc = diaInTractPatch.sum()
+            nDiaSrc = diaInPatch.sum()
 
             self.log.info(
                 "Read DiaSource catalog of length %i from visit %i, "
                 "detector %i. Found %i sources within the patch/tract "
                 "footprint, including %i associated with SSOs.",
-                len(diaCat), diaCatRef.dataId["visit"],
+                nDiaSrcIn, diaCatRef.dataId["visit"],
                 diaCatRef.dataId["detector"], nDiaSrc + nSsSrc, nSsSrc)
 
             if nDiaSrc > 0:
-                diaSourceHistory.append(diaCat)
+                diaSourceHistory.append(diaCat[diaInPatch])
             if nSsSrc > 0:
-                ssSourceHistory.append(associatedSsSources)
-                diaSourceHistory.append(associatedSsDiaSources.to_pandas())
+                ssSourceHistory.append(ssoAssocResult.associatedSsSources)
             if nSsObj > 0:
-                unassociatedSsObjects['visit'] = visit
-                unassociatedSsObjects['detector'] = detector
-                unassociatedSsObjectHistory.append(unassociatedSsObjects)
+                unassociatedSsObjectHistory.append(ssoAssocResult.unassociatedSsObjects)
 
-        if diaSourceHistory:
-            diaSourceHistoryCat = pd.concat(diaSourceHistory)
-        else:
-            diaSourceHistoryCat = diaCat.drop(diaCat.index)
+        # After looping over all of the detector-level catalogs that overlap the
+        # patch, combine them into patch-level catalogs
+        diaSourceHistoryCat = self._stackCatalogs(diaSourceHistory)
         if self.config.doSolarSystemAssociation:
-            nSsSrc, nSsObj = 0, 0
-            if ssSourceHistory:
-                ssSourceHistoryCat = tb.vstack(ssSourceHistory)
-                ssSourceHistoryCat.remove_columns(['ra', 'dec'])
-                nSsSrc = len(ssSourceHistoryCat)
-            else:
-                ssSourceHistoryCat = associatedSsSources  # Empty table?
-            if unassociatedSsObjectHistory:
-                unassociatedSsObjectHistoryCat = tb.vstack(unassociatedSsObjectHistory)
-                nSsObj = len(unassociatedSsObjectHistoryCat)
-            else:
-                unassociatedSsObjectHistoryCat = unassociatedSsObjects  # Empty table?
+            ssSourceHistoryCat = self._stackCatalogs(ssSourceHistory, remove_columns=['ra', 'dec'])
+            nSsSrcTotal = len(ssSourceHistoryCat) if ssSourceHistoryCat else 0
+            unassociatedSsObjectHistoryCat = self._stackCatalogs(unassociatedSsObjectHistory)
+            nSsObjTotal = len(unassociatedSsObjectHistoryCat) if unassociatedSsObjectHistoryCat else 0
```

Comment:
These checks are different than the code block below `if (not diaSourceHistory) and not (self.config.doSolarSystemAssociation and ssSourceHistory)` and could lead to inconsistent behavior in edge cases. It might be wise to cleanup this whole section and have a consistent definition of what it means to have work found.